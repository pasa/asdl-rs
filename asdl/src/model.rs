#[derive(Debug)]
pub struct Asdl {
    pub types: Vec<Type>,
    pub comments: Vec<String>,
}

impl Asdl {
    pub fn get_type_by_name(&self, name: &str) -> Option<&Type> {
        self.types.iter().find(|t| t.id() == name)
    }
}

#[derive(Debug)]
pub enum Type {
    SumType(SumType),
    ProdType(ProdType),
}

impl Type {
    fn id(&self) -> String {
        match self {
            Type::SumType(sty) => sty.id.clone(),
            Type::ProdType(pty) => pty.id.clone(),
        }
    }
}

impl From<SumType> for Type {
    fn from(n: SumType) -> Type {
        Type::SumType(n)
    }
}
impl From<ProdType> for Type {
    fn from(n: ProdType) -> Type {
        Type::ProdType(n)
    }
}

#[derive(Debug)]
pub struct SumType {
    pub id: String,
    pub constructors: Vec<Constructor>,
    pub attributes: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Debug)]
pub struct Constructor {
    pub id: String,
    pub fields: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Debug)]
pub struct ProdType {
    pub id: String,
    pub fields: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Debug)]
pub struct Field {
    pub id: String,
    pub type_id: String,
    pub arity: Arity,
}

#[derive(Debug)]
pub enum Arity {
    Optional,
    Required,
    Repeated,
}
