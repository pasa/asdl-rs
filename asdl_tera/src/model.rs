use std::collections::HashMap;
use serde::Serialize;
use asdl;
use super::Result;

#[derive(Serialize, Debug)]
pub struct Asdl {
    pub types: HashMap<String, Type>,
    pub prod_types: Vec<String>,
    pub sum_types: Vec<String>,
    pub comments: Vec<String>,
}

impl Asdl {
    pub fn parse(asdl: &str) -> Result<Asdl> {
        Ok(Asdl::new(asdl::Asdl::parse(asdl)?))
    }
}

#[derive(Serialize, Debug)]
#[serde(untagged)]
pub enum Type {
    SumType(SumType),
    ProdType(ProdType),
}

impl From<SumType> for Type {
    fn from(n: SumType) -> Type {
        Type::SumType(n)
    }
}
impl From<ProdType> for Type {
    fn from(n: ProdType) -> Type {
        Type::ProdType(n)
    }
}

#[derive(Serialize, Debug)]
pub struct SumType {
    pub id: String,
    pub constructors: Vec<Constructor>,
    pub attributes: Vec<Field>,
    pub is_prod_type: bool, //always false
    pub comments: Vec<String>,
}

#[derive(Serialize, Debug)]
pub struct Constructor {
    pub id: String,
    pub fields: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Serialize, Debug)]
pub struct ProdType {
    pub id: String,
    pub fields: Vec<Field>,
    pub is_prod_type: bool, //always true
    pub comments: Vec<String>,
}

#[derive(Serialize, Debug)]
pub struct Field {
    pub id: String,
    pub type_id: String,
    pub is_required: bool,
    pub is_optional: bool,
    pub is_repeated: bool,
}
