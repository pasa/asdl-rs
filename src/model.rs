use std::convert::From;
use rustc_hash::{FxHashMap, FxHashSet};
use std::fmt::Write;
use heck::{CamelCase, ShoutySnakeCase, SnakeCase, MixedCase};

use serde::{
        Serialize,
};

use crate:: {
        Root,
        TypeKind::*
};

#[derive(Serialize, Debug)]
pub(crate) struct Asdl {
    sum_types: Vec<SumType>,
    prod_types: Vec<ProdType>,
}

#[derive(Serialize, Debug)]
pub(crate) struct SumType {
    id: String,
    constructors: Vec<Constructor>,
}

#[derive(Serialize, Debug)]
pub(crate) struct Constructor {
    id: String,
    prod_type: ProdType,
}

#[derive(Serialize, Debug, Clone, Hash, Eq, PartialEq)]
pub(crate) struct ProdType {
    id: String,
    fields: Vec<Field>,
}

#[derive(Serialize, Debug, Clone, Hash, Eq, PartialEq)]
pub(crate) struct Field {
    id: String,
    type_id: String,
    is_single: bool,
    is_option: bool,
    is_sequence: bool,
}

impl Field {
    fn single(id: String, type_id: String) -> Field {
        Field { id, type_id, is_single: true, is_option: false, is_sequence: false }
    }

    fn option(id: String, type_id: String) -> Field {
        Field { id, type_id, is_single: false, is_option: true, is_sequence: false }
    }

    fn sequence(id: String, type_id: String) -> Field {
        Field { id, type_id, is_single: false, is_option: false, is_sequence: true }
    }
}

impl From<&Root> for Asdl {
    fn from(root: &Root) -> Self {
        let mut res = Asdl { prod_types: vec![], sum_types: vec![] };
        for d in root.defs() {
            let type_id = d.type_id().text().to_string();
            match d.ty().kind() {
                Sum(t) => {
                    res.sum_types.push(sum_type(type_id, t));
                }
                Prod(t) => {
                    res.prod_types.push(prod_type(type_id, t));
                }
            }
        }
        let syntetic_prod_types: FxHashSet<ProdType> = res
            .sum_types
            .iter()
            .flat_map(|t| t.constructors.iter().map(|c| c.prod_type.clone()))
            .collect();
        res.prod_types.extend(syntetic_prod_types);
        res
    }
}

fn sum_type(id: String, node: &crate::SumType) -> SumType {
    let constructors = node.constructors().map(constructor).collect();
    SumType { id, constructors }
}

fn constructor(node: &crate::Constr) -> Constructor {
    let id = node.id().text().to_string();
    let fields = match node.fields() {
        Option::Some(fs) => fields(fs),
        Option::None => vec![],
    };
    let pt = ProdType { id: id.to_mixed_case(), fields };
    Constructor { id, prod_type: pt }
}

fn fields(fields: &crate::Fields) -> Vec<Field> {
    let mut names = FieldNames::default();
    fields.fields().map(|f| field(f, &mut names)).collect()
}

fn field(node: &crate::Field, names: &mut FieldNames) -> Field {
    let type_id = node.type_id().text().to_string();
    let id = names.get_or_generate(node);
    if node.star().is_some() {
        Field::sequence(id, type_id)
    } else if node.q_mark().is_some() {
        Field::option(id, type_id)
    } else {
        Field::single(id, type_id)
    }
}

fn prod_type(type_id: String, node: &crate::ProdType) -> ProdType {
    ProdType { id: type_id, fields: fields(node.fields()) }
}

#[derive(Default)]
struct FieldNames {
    names_indexes: FxHashMap<String, u32>,
}

impl FieldNames {
    fn get_or_generate(&mut self, f: &crate::Field) -> String {
        match f.id() {
            Option::Some(id) => id.text().to_string(),
            Option::None => {
                let type_id = f.type_id().text();
                let index = self.names_indexes.entry(type_id.to_string()).or_insert(0);
                let res = if *index == 0 {
                    type_id.to_string()
                } else {
                    let mut buf = String::new();
                    write!(&mut buf, "{}{}", type_id, index).unwrap();
                    buf
                };
                *index += 1;
                res
            }
        }
    }
}
