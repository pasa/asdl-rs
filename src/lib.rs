mod model;
mod parser;

use tera::*;
use crate::model::*;
use failure::format_err;
use heck::{CamelCase, ShoutySnakeCase, SnakeCase, MixedCase};

pub fn generate(asdl: &str, template: &str) -> String {
    let root: &parser::Root = &parser::parse(&asdl);
    let model = Asdl::from(root);
    let mut tera = Tera::default();
    tera.register_filter("camel", |arg, _| Ok(arg.as_str().unwrap().to_camel_case().into()));
    tera.register_filter("snake", |arg, _| Ok(arg.as_str().unwrap().to_snake_case().into()));
    tera.register_filter("mixed", |arg, _| Ok(arg.as_str().unwrap().to_mixed_case().into()));
    tera.register_filter("SCREAM", |arg, _| {
        Ok(arg.as_str().unwrap().to_shouty_snake_case().into())
    });
    tera.add_raw_template("_src", &template)
        .map_err(|e| format_err!("template parsing error: {:?}", e))
        .unwrap();
    tera.render("_src", &model)
        .map_err(|e| format_err!("template rendering error: {:?}", e))
        .unwrap()
}

#[cfg(test)]
mod tests {
    use crate::*;
    use insta::assert_snapshot_matches;
    use insta::assert_debug_snapshot_matches;

    #[test]
    fn simple_successful_test() {
        let asdl = r"
            stm = Compound(stm s1, stm* s2)
                | Single(stm)
            noFileds = One | Two | Tree
            prodType = (noFileds? f, stm s1)
            ";
        let root: &parser::Root = &parser::parse(&asdl);
        assert_snapshot_matches!("simple_successful_test_syntax", root.debug_dump());
        let model = Asdl::from(root);
        assert_debug_snapshot_matches!("simple_successful_test_model", model)
    }
}
