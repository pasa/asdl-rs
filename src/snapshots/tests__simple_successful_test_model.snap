---
created: "2019-06-17T11:47:23.384116309Z"
creator: insta@0.8.1
source: src/lib.rs
expression: model
---
Asdl {
    sum_types: [
        SumType {
            id: "stm",
            constructors: [
                Constructor {
                    id: "Compound",
                    prod_type: ProdType {
                        id: "compound",
                        fields: [
                            Field {
                                id: "s1",
                                type_id: "stm",
                                is_single: true,
                                is_option: false,
                                is_sequence: false
                            },
                            Field {
                                id: "s2",
                                type_id: "stm",
                                is_single: false,
                                is_option: false,
                                is_sequence: true
                            }
                        ]
                    }
                },
                Constructor {
                    id: "Single",
                    prod_type: ProdType {
                        id: "single",
                        fields: [
                            Field {
                                id: "stm1",
                                type_id: "stm",
                                is_single: true,
                                is_option: false,
                                is_sequence: false
                            }
                        ]
                    }
                }
            ]
        },
        SumType {
            id: "noFileds",
            constructors: [
                Constructor {
                    id: "One",
                    prod_type: ProdType {
                        id: "one",
                        fields: []
                    }
                },
                Constructor {
                    id: "Two",
                    prod_type: ProdType {
                        id: "two",
                        fields: []
                    }
                },
                Constructor {
                    id: "Tree",
                    prod_type: ProdType {
                        id: "tree",
                        fields: []
                    }
                }
            ]
        }
    ],
    prod_types: [
        ProdType {
            id: "prodType",
            fields: [
                Field {
                    id: "f",
                    type_id: "noFileds",
                    is_single: false,
                    is_option: true,
                    is_sequence: false
                },
                Field {
                    id: "s1",
                    type_id: "stm",
                    is_single: true,
                    is_option: false,
                    is_sequence: false
                }
            ]
        },
        ProdType {
            id: "tree",
            fields: []
        },
        ProdType {
            id: "single",
            fields: [
                Field {
                    id: "stm1",
                    type_id: "stm",
                    is_single: true,
                    is_option: false,
                    is_sequence: false
                }
            ]
        },
        ProdType {
            id: "compound",
            fields: [
                Field {
                    id: "s1",
                    type_id: "stm",
                    is_single: true,
                    is_option: false,
                    is_sequence: false
                },
                Field {
                    id: "s2",
                    type_id: "stm",
                    is_single: false,
                    is_option: false,
                    is_sequence: true
                }
            ]
        },
        ProdType {
            id: "one",
            fields: []
        },
        ProdType {
            id: "two",
            fields: []
        }
    ]
}
